lib/irq.c

```diff
@@ -228,162 +235,122 @@ dev_set_irqs_validate(vfu_ctx_t *vfu_ctx, struct vfio_irq_set *irq_set)
     if (irq_set->index >= VFU_DEV_NUM_IRQS) {
-        vfu_log(vfu_ctx, LOG_DEBUG, "bad IRQ index %d\n", irq_set->index);
-        return -EINVAL;
+        line = __LINE__;
+        goto invalid;
     }
 
-    /* TODO make each condition a function */
-
     // Only one of MASK/UNMASK/TRIGGER is valid.
     if ((a_type != VFIO_IRQ_SET_ACTION_MASK) &&
         (a_type != VFIO_IRQ_SET_ACTION_UNMASK) &&
         (a_type != VFIO_IRQ_SET_ACTION_TRIGGER)) {
-        vfu_log(vfu_ctx, LOG_DEBUG, "bad IRQ action mask %d\n", a_type);
-        return -EINVAL;
+        line = __LINE__;
+        goto invalid;
     }
     // Only one of NONE/BOOL/EVENTFD is valid.
     if ((d_type != VFIO_IRQ_SET_DATA_NONE) &&
         (d_type != VFIO_IRQ_SET_DATA_BOOL) &&
         (d_type != VFIO_IRQ_SET_DATA_EVENTFD)) {
-        vfu_log(vfu_ctx, LOG_DEBUG, "bad IRQ data %d\n", d_type);
-        return -EINVAL;
+        line = __LINE__;
+        goto invalid;
     }
     // Ensure irq_set's start and count are within bounds.
     if ((irq_set->start >= vfu_ctx->irq_count[irq_set->index]) ||
         (irq_set->start + irq_set->count > vfu_ctx->irq_count[irq_set->index])) {
-        vfu_log(vfu_ctx, LOG_DEBUG, "bad IRQ start/count\n");
-        return -EINVAL;
+        line = __LINE__;
+        goto invalid;
     }
     // Only TRIGGER is valid for ERR/REQ.
     if (((irq_set->index == VFIO_PCI_ERR_IRQ_INDEX) ||
          (irq_set->index == VFIO_PCI_REQ_IRQ_INDEX)) &&
         (a_type != VFIO_IRQ_SET_ACTION_TRIGGER)) {
-        vfu_log(vfu_ctx, LOG_DEBUG, "bad IRQ trigger w/o ERR/REQ\n");
-        return -EINVAL;
+        line = __LINE__;
+        goto invalid;
+    }
+    // if count == 0, start must be 0 too
+    if ((irq_set->count == 0) && (irq_set->start != 0)) {
+        line = __LINE__;
+        goto invalid;
     }
     // count == 0 is only valid with ACTION_TRIGGER and DATA_NONE.
     if ((irq_set->count == 0) && ((a_type != VFIO_IRQ_SET_ACTION_TRIGGER) ||
                                   (d_type != VFIO_IRQ_SET_DATA_NONE))) {
-        vfu_log(vfu_ctx, LOG_DEBUG, "bad IRQ count %d\n", irq_set->count);
-        return -EINVAL;
+        line = __LINE__;
+        goto invalid;
     }
-    // If IRQs are set, ensure index matches what's enabled for the device.
-    if ((irq_set->count != 0) && (vfu_ctx->irqs->type != IRQ_NONE) &&
-        (irq_set->index != LM2VFIO_IRQT(vfu_ctx->irqs->type))) {
-        vfu_log(vfu_ctx, LOG_DEBUG, "bad IRQ index\n");
-        return -EINVAL;
+    // If fd's are provided, ensure it's only for VFIO_IRQ_SET_DATA_EVENTFD
+    if (nr_fds != 0 && d_type != VFIO_IRQ_SET_DATA_EVENTFD) {
+        line = __LINE__;
+        goto invalid;
+    }
+    // If fd's are provided, ensure they match ->count
+    if (nr_fds != 0 && nr_fds != irq_set->count) {
+        line = __LINE__;
+        goto invalid;
     }
 
     return 0;
+
+invalid:
+    vfu_log(vfu_ctx, LOG_DEBUG, "invalid SET_IRQS (%d): action=%u data_type=%u "
```

Comment:
The line trick is definitely cool but OTOH it's not dead simple for the user to understand why it failed.

Reply:
They can start at the error message too, this is just to make it easier in context.